specs/APIReview_CustomDownload.md

```diff
@@ -0,0 +1,369 @@
+// Selecting `OK` will save the download to the chosen path.
+// Selecting `CANCEL` will cancel the download.
+m_webViewStaging = m_webView.query<ICoreWebView2Staging2>();
+m_demoUri = L"https://demo.smartscreen.msft.net/";
+    CHECK_FAILURE(m_webViewStaging->add_DownloadStarting(
+        Callback<ICoreWebView2StagingDownloadStartingEventHandler>(
+            [this](
+                ICoreWebView2Staging2* sender,
+                ICoreWebView2StagingDownloadStartingEventArgs* args) -> HRESULT {
+                auto showDialog = [this, args] {
+                    // Hide the default download dialog.
+                    CHECK_FAILURE(args->put_ShouldDisplayDefaultDownloadDialog(FALSE));
+
+                    UINT64 downloadSizeInBytes = 0;
+                    CHECK_FAILURE(args->get_DownloadSizeInBytes(&downloadSizeInBytes));
+
+                    wil::unique_cotaskmem_string uri;
+                    CHECK_FAILURE(args->get_Uri(&uri));
+
+                    wil::unique_cotaskmem_string mimeType;
+                    CHECK_FAILURE(args->get_MimeType(&mimeType));
+
+                    wil::unique_cotaskmem_string contentDisposition;
+                    CHECK_FAILURE(args->get_ContentDisposition(&contentDisposition));
+
+                    wil::unique_cotaskmem_string resultFilePath;
+                    CHECK_FAILURE(args->get_ResultFilePath(&resultFilePath));
+
+                    std::wstring prompt =
+                        std::wstring(L"Enter result file path or select `OK` to use default path. ") +
+                        L"Select `Cancel` to cancel the download.";
+
+                    std::wstring description = std::wstring(L"URI: ") + uri.get() + L"\r\n" +
+                                               L"Mime type: " + mimeType.get() + L"\r\n" +
+                                               L"Download size in bytes: " + std::to_wstring(downloadSizeInBytes) +
+                                               L"\r\n";
+
+                    TextInputDialog dialog(
+                        m_appWindow->GetMainWindow(), L"Download Starting", prompt.c_str(),
+                        description.c_str(), resultFilePath.get());
+                    if (dialog.confirmed)
+                    {
+                        // If user selects `OK`, the download will complete normally.
+                        // Result file path will be updated if a new one was provided.
+                        CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
+                    }
+                    else
+                    {
+                        // If user selects `Cancel`, the download will be canceled.
+                        CHECK_FAILURE(args->put_Cancel(TRUE));
+                    }
+                };
+
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
+                CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                m_completeDefferedDownloadEvent = [showDialog, deferral] {
```

Comment:
Maybe a comment explaining how m_completeDefferedDownloadEvent is used and why would be good.